### PR TITLE
Fix multiple issues

### PR DESCRIPTION
**Pull request by ptoews**
Here are my changes so far. As I already mentioned, I've rewritten the receiving by taking the whole datagram from the socket and then iteratively processing that message while removing the processed bytes from the message stored as list (since bytes are immutable), similar to the previous iterative receiving, so that the required changes are not that big. 
I also renamed a few methods to fit better to the new approach and because I actually expected more necessary changes initially.

However I noticed that the entry list is not properly received, so this is still WIP.

Edit: About the entry list receiving: When running the program, it keeps requesting the entry list on every received Realtime_Car_Update, but the response is always another Realtime_Car_Update. If I disable processing of the Realtime_Updates, the program receives the entry list responses. So it seems that since the game sends so many Realtime_Updates, the datagrams regarding the entry list are discarded while the program processes the other messages.

### CODE REVIEW

**Comment by ptoews** [comment]
Lots of more changes:
- The nested list comprehension for the `TrackData` constructor is executed in a different order on python <= 3.7 which causes parsing issues and exceptions. Unwrapped it as a fix (f5b4206)
- I added a timeout state during which the receive loop continues, so that e.g. when the game is in Esc-menu, the api doesn't exit and waits for continuation (b26fd7b)
- Previously the connection request happened before the receiving thread wasn't even started, so I moved that inside the thread to prevent possible issues (f8081ba)
- I also reverted 5f29032 since that didn't actually fix the issue mentioned above. I noticed that sometimes the EntryListCars are sent and received after some time of only receiving RealtimeUpdates.
- Finally, I fixed a big issue I had before, namely that the wrapper could not reconnect (or only after a long delay) if I closed it before one or two times. This happens even with the official demo client, but I found a fix in the ACC forums (https://www.assettocorsa.net/forum/index.php?threads/lets-talk-about-broadcasting-programmers-thread.53826/page-6#post-1137208). However it is important that the `stop()` method is now called when the wrapper exists, which doesn't happen in my PyCharm IDE on Windows, since it seems to just kill the process without any signals. I wrote a quick GUI with a single Quit button for that, and it works super reliably now.

I will continue to check if everything runs stable but I think it should be good now.

File: src/accapi/client.py [hunk 1]
```diff
@@ -51,9 +52,11 @@ def __init__(
         self._onBroadcastingEvent = Observable()
 
         # Session properties
+        self._udpBufferSize = 1024
```
Comment (EmperorCookie) [inline src/accapi/client.py:55]: Buffer has been implemented in a way that follow the original concepts of this repository in #4.

File: src/accapi/client.py [hunk 2]
```diff
@@ -122,36 +125,24 @@ def _send(self, *fmtValuePairs):
                 values.append(v)
         self._socket.sendto(struct.pack(fmt, *values), self._server)
 
-    def _receive(self, fmt):
```
Comment (EmperorCookie) [inline src/accapi/client.py:125]: See #4 for how this was fixed.

File: src/accapi/client.py [hunk 3]
```diff
@@ -122,36 +125,24 @@ def _send(self, *fmtValuePairs):
...
+            return
+        self._connectionId = result.connectionId
         self._update_connection_state("established")
         self._request_entry_list()
         self._request_track_data()
 
-    def _receive_realtime_update(self):
-        args = RealtimeUpdate.receive_args(self._receive)
+    def _receive_realtime_update(self, msg):
+        args = RealtimeUpdate.parse(msg)
```
Comment (EmperorCookie) [inline src/accapi/client.py:139]: No longer necessary with #4 fixes.

File: src/accapi/client.py [hunk 4]
```diff
@@ -160,48 +151,47 @@ def _receive_realtime_car_udpate(self):
...
             callback(Event(self, event = event))
 
-    def _request_connection(
-        self,
-        displayName: str,
-        password: str,
-        updateIntervalMs: int,
-        commandPassword: str
-    ):
+    def _request_connection(self):
```
Comment (EmperorCookie) [inline src/accapi/client.py:178]: In #4, _request_connection now uses self attributes as much as possible, only leaving out passwords so they never get stored.

File: src/accapi/client.py [hunk 5]
```diff
@@ -160,48 +151,47 @@ def _receive_realtime_car_udpate(self):
...
+            ("s", self._commandPassword),
+            # This padding seems to be needed, otherwise the game crashes (especially in multiplayer)
+            ("L", 0)
         )
 
     def _request_disconnection(self):
-        self._send(("B", OutboundMessageTypes.UNREGISTER_COMMAND_APPLICATION.value))
+        self._send(
+            ("B", OutboundMessageTypes.UNREGISTER_COMMAND_APPLICATION.value),
+            ("i", self._connectionId)
```
Comment (EmperorCookie) [inline src/accapi/client.py:193]: I have not found documentation that says this is necessary, but I did add it to #4 as is. I have not had the opportunity to properly test the outcome of this request.

File: src/accapi/client.py [hunk 6]
```diff
@@ -276,15 +266,27 @@ def request_hud_page(self, pageName: str):
...
-                self._update_connection_state("lost")
-                break
-            self._receiveMethods[messageType]()
-        self._socket.close()
-        self._socket = None
+        self._request_connection()
+        try:
+            while not self._stopSignal:
+                try:
+                    message = self._socket.recv(self._udpBufferSize)
```
Comment (EmperorCookie) [inline src/accapi/client.py:273]: Ideally, the client would not be responsible for managing the receive buffer. In #4, a special class has been created that manages this buffer, leaving the client to do exactly what it used to do.

File: src/accapi/client.py [hunk 7]
```diff
@@ -276,15 +266,27 @@ def request_hud_page(self, pageName: str):
...
+                except ConnectionResetError:
+                    self._update_connection_state("lost")
+                    break
+                # If we received a message after a timeout and we were already registered, change connection state
+                if self._connectionState == "timeout" and self._connectionId is not None:
+                    self._update_connection_state("established")
+                messageType = struct.unpack("B", message[:1])[0]
+                self._receiveMethods[messageType](list(message[1:]))
+        finally:
+            if self._connectionId is not None:
```
Comment (EmperorCookie) [inline src/accapi/client.py:286]: There are more things than just a missing connection ID that can go wrong here, in #4, this is now a try/except inside of the finally.

File: src/accapi/client.py [hunk 8]
```diff
@@ -299,6 +301,8 @@ def start(
         password: str,
         commandPassword: str = ""
     ):
+        self._password = password
```
Comment (EmperorCookie) [inline src/accapi/client.py:304]: Passwords should never be stored, especially not in plain text.

File: src/accapi/client.py [hunk 9]
```diff
@@ -307,14 +311,7 @@ def start(
         self._socket.settimeout(1)
         self._connectionId = None
         self._writable = False
-        self._request_connection(
```
Comment (EmperorCookie) [inline src/accapi/client.py:310]: I have not found the rational behind requiring that the _request_connection be executed inside the client thread. This was omitted from #4, please comment on that PR if there's a good reason to move it.

File: src/accapi/structs.py [hunk 10]
```diff
@@ -1,7 +1,29 @@
+import struct
+
 from .enums import *
 
 _MAX_INT = 2 ** 31 - 1
 
+
+def unpack_next(msg: list, fmt):
```
Comment (EmperorCookie) [inline src/accapi/structs.py:8]: Ideally, this kind of code would reside in the client. #4 addresses this issue with ThreadedSocketReader.

File: src/accapi/structs.py [hunk 11]
```diff
@@ -29,8 +51,8 @@ def __init__(self, *args):
         self.sessionIndex = args.pop(0)
         self.sessionType = SESSION_TYPE[args.pop(0)]
         self.sessionPhase = SESSION_PHASE[args.pop(0)]
-        self.sessionTime = args.pop(0) * 1000
-        self.sessionEndTime = args.pop(0) * 1000
+        self.sessionTime = args.pop(0)
```
Comment (EmperorCookie) [inline src/accapi/structs.py:54]: This was a silly math mistake. The goal was to have all times be represented in seconds, therefor it should be a division by 1000. This is fixed in #4, and only one exceptional case remains stored in milliseconds, the update interval.

File: src/accapi/structs.py [hunk 12]
```diff
@@ -201,39 +229,46 @@ def __init__(self, *args):
         self.trackName = args.pop(0)
         self.trackId = args.pop(0)
         self.trackMeters = args.pop(0)
-        self.cameraSets = {args.pop(0): [args.pop(0) for _ in range(args.pop(0))] for _ in range(args.pop(0))}
+        d = {}
```
Comment (EmperorCookie) [inline src/accapi/structs.py:232]: Points granted for readability, not sure which backwards compatibility feature is broken here however. In any case, I've made this even more readable (and hopefully also backwards compatible) in #4.

**Review by EmperorCookie — COMMENTED** [review]
All in all, good fixes. Some modifications changed the original specification quite a lot, so I picked the less intrusive fixes and refactored the other fixes to better match the original design.

Closing this PR as it is now fully integrated into #4. If there are further changes required, please open a new PR.

Thank you very much for contributing!!

**Pull request closed without merging**